File: src/Systems/C++/example2.cpp
#include <iostream>

// Abstract Product
class Database {
public:
    virtual void connect() = 0;
    virtual void query() = 0;
};

// Concrete Product
class Postgres : public Database {
public:
    void connect() override {
        std::cout << "Connecting to PostgreSQL" << std::endl;
    }
    void query() override {
        std::cout << "Querying PostgreSQL" << std::endl;
    }
};

class MySQL : public Database {
public:
    void connect() override {
        std::cout << "Connecting to MySQL" << std::endl;
    }
    void query() override {
        std::cout << "Querying MySQL" << std::endl;
    }
};

// Abstract Factory
class DatabaseFactory {
public:
    virtual Database* createDatabase() = 0;
};

// Concrete Factory
class PostgresFactory : public DatabaseFactory {
public:
    Database* createDatabase() override {
        return new Postgres();
    }
};

class MySQLFactory : public DatabaseFactory {
public:
    Database* createDatabase() override {
        return new MySQL();
    }
};

// Usage
void useDatabase(DatabaseFactory* factory) {
    Database* db = factory->createDatabase();
    db->connect();
    db->query();
    delete db;
}

int main() {
    DatabaseFactory* postgresFactory = new PostgresFactory();
    DatabaseFactory* mysqlFactory = new MySQLFactory();
    useDatabase(postgresFactory);
    useDatabase(mysqlFactory);
    delete postgresFactory;
    delete mysqlFactory;
    return 0;
}

File: src/Systems/Rust/example2.rs
trait Database {
    fn connect(&self);
    fn query(&self);
}

struct Postgres;

impl Database for Postgres {
    fn connect(&self) {
        println!("Connecting to PostgreSQL");
    }

    fn query(&self) {
        println!("Querying PostgreSQL");
    }
}

struct MySQL;

impl Database for MySQL {
    fn connect(&self) {
        println!("Connecting to MySQL");
    }

    fn query(&self) {
        println!("Querying MySQL");
    }
}

trait DatabaseFactory {
    fn create_database(&self) -> Box<dyn Database>;
}

struct PostgresFactory;

impl DatabaseFactory for PostgresFactory {
    fn create_database(&self) -> Box<dyn Database> {
        Box::new(Postgres)
    }
}

struct MySQLFactory;

impl DatabaseFactory for MySQLFactory {
    fn create_database(&self) -> Box<dyn Database> {
        Box::new(MySQL)
    }
}

fn use_database(factory: &dyn DatabaseFactory) {
    let db = factory.create_database();
    db.connect();
    db.query();
}

fn main() {
    let postgres_factory = PostgresFactory;
    let mysql_factory = MySQLFactory;
    use_database(&postgres_factory);
    use_database(&mysql_factory);
}
